src/core_plugins/kibana/public/dashboard/reducers/embeddables.ts

```diff
@@ -0,0 +1,120 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import _ from 'lodash';
+import { Reducer } from 'redux';
+
+import {
+  EmbeddableActionTypeKeys,
+  EmbeddableErrorActionPayload,
+  EmbeddableIsInitializedActionPayload,
+  EmbeddableIsInitializingActionPayload,
+  PanelActionTypeKeys,
+  SetStagedFilterActionPayload,
+} from '../actions';
+import {
+  EmbeddableReduxState,
+  EmbeddablesMap,
+  PanelId,
+} from '../selectors/types';
+
+const embeddableIsInitializing = (
+  embeddables: EmbeddablesMap,
+  payload: EmbeddableIsInitializingActionPayload
+): EmbeddablesMap => ({
+  ...embeddables,
+  [payload]: {
+    error: undefined,
+    initialized: false,
+    metadata: {},
+    stagedFilter: undefined,
+  },
+});
+
+const embeddableIsInitialized = (
+  embeddables: EmbeddablesMap,
+  payload: EmbeddableIsInitializedActionPayload
+): EmbeddablesMap => ({
+  ...embeddables,
+  [payload.panelId]: {
+    ...embeddables[payload.panelId],
+    initialized: true,
+    metadata: { ...payload.metadata },
+  },
+});
+
+const setStagedFilter = (
+  embeddables: EmbeddablesMap,
+  payload: SetStagedFilterActionPayload
+): EmbeddablesMap => ({
+  ...embeddables,
+  [payload.panelId]: {
+    ...embeddables[payload.panelId],
+    stagedFilter: payload.stagedFilter,
+  },
+});
+
+const embeddableError = (
+  embeddables: EmbeddablesMap,
+  payload: EmbeddableErrorActionPayload
+): EmbeddablesMap => ({
+  ...embeddables,
+  [payload.panelId]: {
+    ...embeddables[payload.panelId],
+    error: payload.error,
+  },
+});
+
+const clearStagedFilters = (embeddables: EmbeddablesMap): EmbeddablesMap => {
+  const omitStagedFilters = (
+    embeddable: EmbeddableReduxState
+  ): EmbeddablesMap => _.omit({ ...embeddable }, ['stagedFilter']);
+  return _.mapValues<EmbeddablesMap>(embeddables, omitStagedFilters);
+};
+
+const deleteEmbeddable = (
+  embeddables: EmbeddablesMap,
+  panelId: PanelId
+): EmbeddablesMap => {
+  const embeddablesCopy = { ...embeddables };
+  delete embeddablesCopy[panelId];
+  return embeddablesCopy;
+};
+
+export const embeddablesReducer: Reducer<EmbeddablesMap> = (
+  embeddables: EmbeddablesMap = {},
+  action
+): EmbeddablesMap => {
+  switch (action.type) {
+    case EmbeddableActionTypeKeys.EMBEDDABLE_IS_INITIALIZING:
+      return embeddableIsInitializing(embeddables, action.payload);
+    case EmbeddableActionTypeKeys.EMBEDDABLE_IS_INITIALIZED:
+      return embeddableIsInitialized(embeddables, action.payload);
+    case EmbeddableActionTypeKeys.SET_STAGED_FILTER:
+      return setStagedFilter(embeddables, action.payload);
+    case EmbeddableActionTypeKeys.CLEAR_STAGED_FILTERS:
+      return clearStagedFilters(embeddables);
+    case EmbeddableActionTypeKeys.EMBEDDABLE_ERROR:
+      return embeddableError(embeddables, action.payload);
+    case PanelActionTypeKeys.DELETE_PANEL:
```

Comment:
Having PanelActionTypeKeys here is a code smell to me, but I'm not sure of the best way to fix it, and keep a de-normalized data structure.  For this PR at least, I'll leave it as it was before, rather than make any implementation changes.  But for the record, I don't like it (and it was an easy mistake to make - I accidentally left it off the first time, but luckily tests caught the issue).

Reply:
I don't think it's too much of a code smell w/ Redux apps. It's kind of the way things are. I've seen some codebases change their actions to be more generic if they know that more than one handler will need to react to them. But honestly, with this being TypeScripted, it'll be easy to detect all the places affected by a change to an action, which makes this fairly easy to maintain. And that, to me, is the biggest concern w/ leaking actions across boundaries.